### PR TITLE
 Enable Plugins pages for logged-out users

### PR DESCRIPTION
**Pull request by claudiucelfilip**
#### Proposed Changes
More info: pau2Xa-4j7-p2

This allows logged-out users to visit the following Plugin pages:
* The /plugins landing page
* The plugins detail pages – eg. /plugins/mailpoet
* The plugins category pages – eg. /plugins/browse/booking
* Plugins search result pages – eg. /plugins?s=commerce & /plugins/browse/design?s=ecommerce

Changes for this include:
* adjusting the section config to support logged-out routes
* adds data-less SSR routes
* enable localization for logged-out instances by using the URL param (ex. /es/plugins/mailpoet)
* hide current site-specific sections from logged-out users

#### Testing Instructions
* If you're running this on a locale instance, you'll need to run `yarn build-languages` so that the translations work
* Make sure you're logged out
* Go to `/plugins`
* The search box, categories links, and plugins should load and work properly
Screenshot on 2022-08-22 at 14-31-15

* Click on one of the categories or popular searches
* The plugins should load and the top bread crumb should work as logged in
Screenshot on 2022-08-22 at 14-37-03

* The search box should also work and the breadcrumbs should update
Screenshot on 2022-08-22 at 14-35-57

* Clicking on any of the plugins should show the Plugin Details page
Screenshot on 2022-08-22 at 14-37-36

* Go to `/es/plugins` (or any other supported locale prefix)
* Repeat the steps above
* Make sure the translations work as they do when logged in (NOTE: some plugins are not translated)
* The `/es` locale prefix should be maintained in the URLs while navigating through the pages
Screenshot on 2022-08-22 at 15-52-29
Screenshot on 2022-08-22 at 15-54-02

* all the links that should be inaccessible and should redirect to log in
  * `/plugins/manage`
  * `/plugins/upload`
  * `/plugins/setup`
  * `/plugins/active`
  * `/plugins/inactive`
  * `/plugins/updates`
  * `/plugins/wordpress-seo-premium/eligibility`
  * any link which has a `SITE_ID` in it (ex. `/plugins/browse/seo/[SITE_ID]`)

* log in and visit the same links
* everything should work as before

#### TODOs

- [x] Enable Plugins pages for logged-out users (Client-side render + localized)
- [x] Enable Server-side rendering for Plugins pages (#66344)
- [x] Replace popular search words and categories spans with crawlable links
- [x] Implement the Logged-out Plugin details page CTA pointing to the /pricing page

### CODE REVIEW

File: client/my-sites/plugins/index.web.js
Comment (claudiucelfilip) [inline client/my-sites/plugins/index.web.js:53]: I had to go back on the suggestion to have [separate routes for logged in / logged out users](https://github.com/Automattic/wp-calypso/pull/66087/files#r934127573) because there's a route that can be accessed in both situations: `/plugins`. 
I tried separating based on if there's a site selected (as in /themes) but things got a bit hairy and couldn't find a proper workaround for the `/plugins` route.

Using this approach also addresses an issue where the Navigation CSS was not loaded properly when directly accessing the Plugins Detail page.

File: client/my-sites/plugins/plugin-details.jsx
Comment (claudiucelfilip) [inline client/my-sites/plugins/plugin-details.jsx:151]: `translate.localeSlug` serves as a locale fallback for logged-out users where the URL `lang` parameter is used.

File: client/my-sites/plugins/plugins-browser-item/test/index.jsx
Comment (claudiucelfilip) [inline client/my-sites/plugins/plugins-browser-item/test/index.jsx:17]: Needed to add an empty state object so that the `isUserLoggedIn` selector from `useLocalizedPlugins` doesn't throw an error.

File: client/my-sites/plugins/plugin-details-header/index.jsx
Comment (claudiucelfilip) [inline client/my-sites/plugins/plugin-details-header/index.jsx]: The `/plugin/` is invalid and clicking it didn't work. Changed to `/plugins/`.

File: client/my-sites/plugins/utils.js
Comment (lsl) [inline client/my-sites/plugins/utils.js:30]: Could we move this into `localizeUrl` under `i18n-utils`? i.e. add the code somewhere [around here](https://github.com/Automattic/wp-calypso/blob/888ab78745549c6af6f966a58875fd14d9aea3d5/packages/i18n-utils/src/localize-url.tsx#L153).

It may be a good idea to add something similar to `localize_url` in `locales.php` too. Could @Automattic/i18n help with that at all?

Comment (claudiucelfilip) [reply client/my-sites/plugins/utils.js:30]: > Could we move this into localizeUrl under i18n-utils?

@lsl I think that sounds like a good idea but I'm a bit hesitant to add more changes to this PR. 
Considering `localizeUrl` is used in many places, we should probably create a dedicated PR and also adapt these changes to work with the /themes section where there's a [similar function](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/themes/helpers.js#L54).

Comment (lsl) [reply client/my-sites/plugins/utils.js:30]: Follow up sounds good to me.

File: client/components/fixed-navigation-header/index.tsx
Comment (tyxla) [inline client/components/fixed-navigation-header/index.tsx:16]: TBH I wouldn't expect visual changes in a PR that works with sections, controllers and logged in/out users. Any chance we might have an opportunity to split to multiple PRs?

File: client/my-sites/plugins/plugin-details-header/index.jsx
Comment (tyxla) [inline client/my-sites/plugins/plugin-details-header/index.jsx:130]: The `localizePath()` changes in this PR seem like a good modular piece we could extract into its own PR.

File: client/my-sites/plugins/utils.js
Comment (tyxla) [reply client/my-sites/plugins/utils.js:30]: I would actually like to see this new utility in its own PR, with tests, and with approval from the @Automattic/i18n  team. Then we could use that utility to localize the plugins.

File: client/sections.js
Comment (tyxla) [inline client/sections.js:109]: Now that we're starting to use those in more routes than `themes` ones, I think it's time to generalize them and reuse them, rather than redefine them for all localized routes.

**Review by tyxla — COMMENTED** [review]
Thanks for working on this 👍 

I'd love to try to split this into smaller pieces if possible.

Also, I've noticed that this breaks `/plugins/:site` for Jetpack sites in subdirectories. Note that forward slashes are usually escaped as `::` in slugs. Let's make sure we handle that scenario.

Another thing I've noticed: when loading a plugin page for logged-out users (https://container-great-curie.calypso.live/plugins/wordpress-seo) the placeholders are a bit awkward. Are we planning to improve those? Or should we leverage SSR? 

One more thing: previously, when as a logged-out user we tried to access `/plugins/:site`, we were redirected to the log-in page - this is no longer the case, and I think that's a regression.

I'd strongly suggest doing some tests against production and ensuring we're not introducing any regressions while enabling plugins for the logged-out user.

File: client/my-sites/plugins/utils.js
Comment (claudiucelfilip) [reply client/my-sites/plugins/utils.js:30]: @tyxla @lsl I opened #66528 to add the localization functionality to the `localizeUrl` function.

File: client/components/fixed-navigation-header/index.tsx
Comment (claudiucelfilip) [reply client/components/fixed-navigation-header/index.tsx:16]: Fair enough. I moved the CSS Header fixes to https://github.com/Automattic/wp-calypso/pull/66529.

I should mention that this branch is set as the root branch for all the effort needed for the MVP release of getting the Logged Out + SSR Plugins section.

Comment (tyxla) [reply client/components/fixed-navigation-header/index.tsx:16]: Noting that this workflow makes it difficult to understand what to review and what not. For example, these changes are still in this PR and it's confusing to just "have to ignore them". Also, if we had merged those PRs to trunk already, we could have had the opportunity to discover and fix related bugs. That's why we usually try to keep PRs atomic and merge incrementally, not in a single large PR. We can always use feature flags to ease our job if there are incremental changes that will be visible in production.

File: client/my-sites/plugins/plugin-details-header/index.jsx
Comment (claudiucelfilip) [reply client/my-sites/plugins/plugin-details-header/index.jsx:130]: Opened #66528 for this but I don't want it to block this. I suggest we go ahead with this for now and follow up afterward with a PR to replace it here and also in the /themes section.

File: client/sections.js
Comment (claudiucelfilip) [reply client/sections.js:109]: We could probably also use the magnificent locales from the start here and maybe avoid some unnecessary checks down the pipeline.

I'd also suggest we not block the PR with this, and address this in a follow-up.

**Review by worldomonation — APPROVED** [review]
Approved from @Automattic/kitkat end concerning changes under `test/e2e`.

**Comment by mreishus** [comment]
When I try urls like `http://calypso.localhost:3000/es/plugins` or `http://calypso.localhost:3000/fr/plugins` while logged out, I'm only seeing English on the page:

I'm on commit f43489c44f. Expected to see: At least some words in Spanish (or French) like in the screenshots at the top of the PR.

Any links I click drop the `/es` from the url (Example actual link: `http://calypso.localhost:3000/plugins/wordpress-seo-premium`, expected to see: `http://calypso.localhost:3000/es/plugins/wordpress-seo-premium`).

`http://calypso.localhost:3000/zz/plugins` does 404 as expected, so it feels like it's being read somewhere, but not taking effect.

**Review by niranjan-uma-shankar — COMMENTED** [review]
This tested well for me. The non-EN plugins page did not work (i.e not translated) in my local calypso instance but worked in the calypso.live instance, not sure why. I could confirm I could see the translations on calypso.live. There are a few UI issues:

1. On mobile view, the plugin snippet is getting cut-off. In the example below, the full text that's supposed to be visible is `All the benefits of Yoast SEO Free as well as exclusive SEO features which unlock even more tools and functionality.`
Screenshot 2022-08-26 at 9 32 44 AM

2. In mobile view plugin details page, there's unnecessary white space. My guess is that this PR did not introduce this issue so we can take it as a follow up. 

Screenshot 2022-08-26 at 9 33 03 AM

3. In the plugin detail page, I feel that we need to do more explaining about eligible plans for plugins. If I were a fresh user on the plugin detail page, then clicking the "View Plans" takes me to the `/pricing` page. But it's not clear to me that this plugin can be installed only on the Business plan or higher unless I spend time reading the plan features. Can we mention on the plugin details page side section that one needs to be on the Business plan to install plugins?

4. On the plugin details page, I also found the word "Free" followed by "View Plans" below confusing. If it's free, why should I view plans? 

Screenshot 2022-08-25 at 10 32 37 PM

5. We could maybe take this in a follow-up, but I think the CTA on the logged-out plugin details page should take to a dedicated signup flow which adds the plugin to cart. The plans step of this signup flow should show only two options - Business and eCommerce. The `/themes` page already does this - if you select a theme, then it takes to a dedicated signup flow with the theme added to cart. 

I think only point (1) here is a must-fix, the rest can be taken as improvements post-launch. I approve this PR from my side but I suggest getting one more approval from a team that is closer to the plugins code.

**Comment by claudiucelfilip** [comment]
@mreishus @niranjan-uma-shankar Thanks for the reviews!

I double-checked and the translation should be working on the latest commit [f43489c](http://calypso.localhost:3000/plugins).
Screenshot on 2022-08-26 at 10-21-50

It is possible that you don't have the languages built on your local instance. There should be a `/public/languages` folder with a bunch of json files. These are fetched by the [ssrSetupLocale](https://github.com/Automattic/wp-calypso/blob/trunk/client/controller/ssr-setup-locale.js#L39) middleware.
Running `yarn build-languages` should generate it.

> I think only point (1) here is a must-fix, the rest can be taken as improvements post-launch. I approve this PR from my side but I suggest getting one more approval from a team that is closer to the plugins code.

I agree, the PR is already too big and I think that any non-crititcal fix should be avoided this late in the game. 
I'm also seeing the issue in point (1) on happening on production:
Screenshot on 2022-08-26 at 10-28-21
I agree it's a must-fix but I'd avoid adding a fix to this PR if it's not caused by these changes.

Your points (3) (4) are (5) are valid and have been considered. There's a follow-up task planned to create a dedicated onboarding flow for this CTA as a post-MVP improvement (more info in pau2Xa-4j7-p2). The details haven't been drawn up yet, but your suggestion makes sense.

**Comment by niranjan-uma-shankar** [comment]
> I agree it's a must-fix but I'd avoid adding a fix to this PR if it's not caused by these changes.

Agree, if it's not caused by this PR, let's postpone the fix. 

> There's a follow-up task planned to create a dedicated onboarding flow for this CTA as a post-MVP improvement (more info in pau2Xa-4j7-p2). The details haven't been drawn up yet, but your suggestion makes sense.

Great, looks like I missed reading those parts in the PT. 👍 

LGTM from my side.

**Comment by mreishus** [comment]
> It is possible that you don't have the languages built on your local instance. There should be a `/public/languages` folder with a bunch of json files. These are fetched by the [ssrSetupLocale](https://github.com/Automattic/wp-calypso/blob/trunk/client/controller/ssr-setup-locale.js#L39) middleware. Running `yarn build-languages` should generate it.

Thanks, I also needed to `BUILD_TRANSLATION_CHUNKS=true yarn start` then  `yarn build-languages` was able to build the needed files. No longer seeing any issues in this area. :+1:

**Review by mreishus — APPROVED** [review]
LGTM. 

- I'm able to browse the plugins page while logged out.
  - URLs with locales in them work properly and are preserved on click. 
  - Functionality that requires a site, like upload plugins or manage plugins urls, asks you to log in.
- I don't see any regressions in logged in plugin views.
- The only differences I see are coming from plugin related flags that are enabled on dev only.
- I also noticed the cut-off problem mentioned above, but it's also happening in production, so not unique to this PR.

File: client/my-sites/plugins/plugin-details-header/test/index.jsx
Comment (lsl) [inline client/my-sites/plugins/plugin-details-header/test/index.jsx:15]: **Not a blocker**, is there any reason we're limiting this to mag 16? 

Is it just that it's easier than figuring out translation rates for this code? (Some non mag 16 locales have good translation in calypso)

**Review by lsl — COMMENTED** [review]
Nice work! Mostly LGTM, I don't really want to block this anymore than it has been but a few of the mobile views are pretty broken, especially the localized versions. 

Do we need to ship localized pages on day one? Is it worth breaking this down further into two PR's for logged out and logged out localization? (might not be worth it at this point so please take this as an idea more than a suggestion)

1. Managed to get a localized 'en' route, this would be good to fix before Google gets confused by it.

Screenshot_2022-08-29_16-09-44

Repro: Click a category > click a suggested search term > click the category breadcrumb.

I think we're missing the `localizePath` treatment on [these breadcrumbs](https://github.com/Automattic/wp-calypso/blob/6898fb1522d7b177570de35c0a72f3649a479b25/client/my-sites/plugins/plugins-navigation-header/index.jsx#L115-L144).

2. I think you're aware but logged out pages aren't handling mobile well. Are we linking to this anywhere yet? Might not be a blocker.

Screenshot(7)

3. `/jp/plugins` 404's for me, my translation build is failing, possibly why. If you're able to get it to load, please ignore. Not a blocker but needs fixing.

4. We should use pagination for search results on logged out pages but I think that is in follow up territory, made an issue: https://github.com/Automattic/wp-calypso/issues/67075

File: client/my-sites/plugins/plugin-details-header/test/index.jsx
Comment (claudiucelfilip) [reply client/my-sites/plugins/plugin-details-header/test/index.jsx:15]: The main reason is assuring consistency with the path [localization functionality](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/themes/helpers.js#L55) on /themes which supports mag 16. 
I needed to mock this because the newly added `localizePath` was breaking the component tests.

**Comment by claudiucelfilip** [comment]
@lsl Thanks for the review!
> Do we need to ship localized pages on day one? Is it worth breaking this down further into two PR's for logged out and logged out localization? (might not be worth it at this point so please take this as an idea more than a suggestion)

In retrospect, I would have definitely separated this but it's probably too late for this ATM.

> 1. Managed to get a localized 'en' route, this would be good to fix before Google gets confused by it.
> I think we're missing the localizePath treatment on [these breadcrumbs](https://github.com/Automattic/wp-calypso/blob/6898fb1522d7b177570de35c0a72f3649a479b25/client/my-sites/plugins/plugins-navigation-header/index.jsx#L115-L144).

Thanks for catching this! It must have gotten reverted in one of the rebasing sessions. I'll restore these in the follow-up [PR to enable SSR](https://github.com/Automattic/wp-calypso/pull/66344) to avoid any surprises with this PR.

> 2. I think you're aware but logged out pages aren't handling mobile well. Are we linking to this anywhere yet? Might not be a blocker.

I'm also seeing this on /themes production. It's probably something we should address in both sections in a follow-up PR.

Screenshot on 2022-08-29 at 10-28-16

>  3. /jp/plugins 404's for me, my translation build is failing, possibly why. If you're able to get it to load, please ignore. Not a blocker but needs fixing.

It's 404's but also for /jp/themes production. The correct locale param should be `ja`.

File: client/my-sites/plugins/index.web.js
Comment (lsl) [reply client/my-sites/plugins/index.web.js:53]: Didn't take a close enough look apparently, a lot of these are localized but shouldn't be. Logged out users don't need access to most of these logged in only routes. Doesn't seem to be any upside to having these localized, is there any downsides to leaving them as is?

Made an issue to fix, let me know if there are any issues with that. https://github.com/Automattic/wp-calypso/issues/67711

Comment (claudiucelfilip) [reply client/my-sites/plugins/index.web.js:53]: >  Doesn't seem to be any upside to having these localized, is there any downsides to leaving them as is?

I agree but we'll still need to keep the `/${ langParam }` and `redirectWithoutLocaleParamIfLoggedIn` for unlocalized routes. 
Otherwise, these routes will end up being caught by the Plugin details route: `/${ langParam }/plugins/:plugin/:site_id?`

To avoid localizing the logged-in routes, we'll need to specify them `index.node.js` file.

Comment (lsl) [reply client/my-sites/plugins/index.web.js:53]: Can you explain how this works a bit more? We just added this route: https://github.com/Automattic/wp-calypso/pull/67710/files?w=1#diff-7edd3c450e9fa353a0acd99fdf2b07d0c902ef5c88f419a3f9ddd3df242822f8R92

Comment (claudiucelfilip) [reply client/my-sites/plugins/index.web.js:53]: For example, while logged-in and accessing a link such as `/ja/plugins/plans/{SITE}`:
* first it gets caught by the plugins details route `/${ langParam }/plugins/:plugin/:site_id?`
* `redirectWithoutLocaleParamIfLoggedIn` redirects to `/plugins/plans/{SITE}`
* `/plugins/plans/:site` now correctly catches the request

Other scenarios such as accessing `/ja/plugins/plans/{SITE}` as logged-out is also caught by the plugin details route but that also redirects to login because of the `{SITE}` in the URL. 
If `{SITE}` is not present, then it'll show an empty/placeholder details page as it can't find the `plans` plugin.

Comment (lsl) [reply client/my-sites/plugins/index.web.js:53]: Ah didn't see langParam was optional or how that redirect was working, gotcha, thanks.

**Pull request merged**
